src/mo_interpreter/interpret.ml

```diff
@@ -885,16 +885,14 @@ and interpret_dec env dec (k : V.value V.cont) =
   match dec.it with
   | ExpD exp ->
     interpret_exp env exp k
-  | LetD (pat, exp, None) ->
-    interpret_exp env exp (fun v ->
-      define_pat env pat v;
-      k v
-    )
-  | LetD (pat, exp, Some fail) -> 
-    interpret_exp env exp (fun v ->
-      let err () = interpret_exp env fail (fun _ -> assert false) in
-      define_pat' env err pat v;
-      k v
+  | LetD (pat, exp, fail) -> 
+    interpret_exp env exp (fun v -> 
+      let err () = 
+        match fail with
+        | Some fail -> interpret_exp env fail (fun _ -> assert false)
+        | None -> trap pat.at "value %s does not match pattern" (string_of_val env v)
+      in
+      define_pat env k err pat v
```

Comment:
This looks like a particularly clean solution to the `return` problem. Great you figured this out!

Reply:
Yay